File: src/main/java/com/chatchatabc/jpademojava/domain/repository/PostRepository.java
package com.chatchatabc.jpademojava.domain.repository;

import com.chatchatabc.jpademojava.domain.model.Post;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.time.Instant;

@Repository
public interface PostRepository extends JpaRepository<Post, String> {

    /**
     * Find all posts between date range
     *
     * @param startDate
     * @param endDate
     * @return
     */
    @Query("SELECT p FROM Post p WHERE p.createdAt BETWEEN :startDate AND :endDate")
    Page<Post> findAllBetweenDateRange(Instant startDate, Instant endDate, Pageable pageable);

    /**
     * Find all posts between date range by user
     *
     * @param startDate
     * @param endDate
     * @param userId
     * @return
     */
    @Query("SELECT p FROM Post p WHERE p.createdAt BETWEEN :startDate AND :endDate AND p.user.id = :userId")
    Page<Post> findAllBetweenDateRangeByUser(Instant startDate, Instant endDate, String userId, Pageable pageable);
}

File: src/main/kotlin/com/chatchatabc/jpademo/domain/repository/PostRepository.kt
package com.chatchatabc.jpademo.domain.repository

import com.chatchatabc.jpademo.domain.model.Post
import com.chatchatabc.jpademo.domain.model.User
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.Instant

@Repository
interface PostRepository : JpaRepository<Post, String> {

    /**
     * Find all posts between date range
     */
    @Query("SELECT p FROM Post p WHERE p.createdAt BETWEEN :startDate AND :endDate")
    fun findAllBetweenDateRange(startDate: Instant?, endDate: Instant?, pageable: Pageable): Page<Post>

    /**
     * Find all posts between date range and by user param optionally if not null.
     */
    @Query("SELECT p FROM Post p WHERE p.user = :user AND p.createdAt BETWEEN :startDate AND :endDate")
    fun findAllBetweenDateRangeByUser(startDate: Instant?, endDate: Instant?, user: User, pageable: Pageable): Page<Post>
}